backend/tests/unit/danswer/llm/answering/stream_processing/test_citation_processing.py

```diff
@@ -0,0 +1,269 @@
+            "[[1]](https://0.com) Citation at the beginning. [[2]]() In the middle. At the end [[3]](https://2.com).",
+            ["doc_0", "doc_1", "doc_2"],
+        ),
+        (
+            "Mixed valid and invalid citations",
+            [
+                "Mixed valid and in",
+                "valid citations ",
+                "[",
+                "1",
+                "]",
+                "[",
+                "99",
+                "]",
+                "[",
+                "3",
+                "]",
+                "[",
+                "100",
+                "]",
+                "[",
+                "5",
+                "]",
+                ".",
+            ],
+            "Mixed valid and invalid citations [[1]](https://0.com)[99][[2]]()[100][[3]](https://2.com).",
+            ["doc_0", "doc_1", "doc_2"],
+        ),
+        (
+            "Hardest!",
+            [
+                "Multiple cit",
+                "ations in one ",
+                "sentence [",
+                "1",
+                "]",
+                "[",
+                "4",
+                "]",
+                "[",
+                "5",
+                "]",
+                ". ",
+            ],
+            "Multiple citations in one sentence [[1]](https://0.com)[[2]]()[[3]](https://2.com).",
+            ["doc_0", "doc_1", "doc_2"],
+        ),
+        (
+            "Repeated citations with text",
+            ["[", "1", "]", "Aasf", "asda", "sff  ", "[", "1", "]", " ."],
+            "[[1]](https://0.com)Aasfasdasff  [[1]](https://0.com) .",
+            ["doc_0"],
+        ),
+        (
+            "Consecutive identical citations!",
+            [
+                "Citations [",
+                "1",
+                "]",
+                "[",
+                "1]",
+                "",
+                "[2",
+                "",
+                "]",
+                ". ",
+            ],
+            "Citations [[1]](https://0.com).",
+            ["doc_0"],
+        ),
+        (
+            "Consecutive identical citations!",
+            [
+                "test [1]tt[1]t",
+                "",
+            ],
+            "test [[1]](https://0.com)ttt",
+            ["doc_0"],
+        ),
+        (
+            "Consecutive identical citations!",
+            [
+                "test [1]t[1]t[1]",
+                "",
+            ],
+            "test [[1]](https://0.com)tt",
+            ["doc_0"],
+        ),
+        (
+            "Repeated citations with text",
+            ["[", "1", "]", "Aasf", "asda", "sff  ", "[", "1", "]", " ."],
+            "[[1]](https://0.com)Aasfasdasff  [[1]](https://0.com) .",
+            ["doc_0"],
+        ),
+        (
+            "Repeated citations with text",
+            ["[1][", "1", "]t", "[2]"],
+            "[[1]](https://0.com)t",
+            ["doc_0"],
+        ),
+        (
+            "Repeated citations with text",
+            ["[1][", "1", "]t]", "[2]"],
```

Comment:
Just to be extra clear, I imagine it works fine but what if ][ is its own token, it may be different since
"[1][" as a token handles a complete citation, where as "[1" + "][" only becomes a complete citation after the second token. Anyway, I'm sure it's fine